onadata/libs/serializers/project_invitation_serializer.py

```diff
@@ -0,0 +1,129 @@
+
+        if domain_whitelist:
+            # Extract the domain from the email address
+            domain = email.split("@")[1]
+
+            # Check if the domain matches "foo.com"
+            if not domain.lower() in [
+                allowed_domain.lower() for allowed_domain in domain_whitelist
+            ]:
+                raise serializers.ValidationError(_(err_msg))
+
+        # email should not be of an existing user
+        if User.objects.filter(email=email).exists():
+            raise serializers.ValidationError(_("User already exists"))
+
+        return email
+
+    def validate_role(self, role):
+        """Validate `role` field"""
+        if role not in ROLES:
+            raise serializers.ValidationError(_("Invalid role."))
+
+        return role
+
+    def get_unique_together_validators(self):
+        """Overriding method to disable unique together checks"""
+        return []
+
+    class Meta:
+        model = ProjectInvitation
+        fields = (
+            "id",
+            "email",
+            "project",
+            "role",
+            "status",
+        )
+        read_only_fields = ("status",)
+        extra_kwargs = {"project": {"write_only": True}}
+
+
+# pylint: disable=abstract-method
+class ProjectInvitationUpdateBaseSerializer(serializers.Serializer):
+    """Base serializer for project invitation updates"""
+
+    invitation_id = serializers.IntegerField()
+
+    def validate_invitation_id(self, invitation_id):
+        """Validate `invitation_id` field"""
+        try:
+            ProjectInvitation.objects.get(pk=invitation_id)
+
+        except ProjectInvitation.DoesNotExist as e:
+            raise serializers.ValidationError(_("Invalid invitation_id.")) from e
+
+        return invitation_id
+
+
+class ProjectInvitationRevokeSerializer(ProjectInvitationUpdateBaseSerializer):
+    """Serializer for revoking a project invitation"""
+
+    def validate_invitation_id(self, invitation_id):
+        super().validate_invitation_id(invitation_id)
+
+        invitation = ProjectInvitation.objects.get(pk=invitation_id)
+
+        if invitation.status != ProjectInvitation.Status.PENDING:
+            raise serializers.ValidationError(
+                _("You cannot revoke an invitation which is not pending")
+            )
+
+        return invitation_id
+
+    def save(self, **kwargs):
+        invitation_id = self.validated_data.get("invitation_id")
+        invitation = ProjectInvitation.objects.get(pk=invitation_id)
+        invitation.status = ProjectInvitation.Status.REVOKED
+        invitation.save()
+
+
+class ProjectInvitationResendSerializer(ProjectInvitationUpdateBaseSerializer):
+    """Serializer for resending a project invitation"""
+
+    def validate_invitation_id(self, invitation_id):
+        super().validate_invitation_id(invitation_id)
+
+        invitation = ProjectInvitation.objects.get(pk=invitation_id)
+
+        if invitation.status != ProjectInvitation.Status.PENDING:
```

Comment:
Should it be possible to resend rejected invitations ?

Reply:
@DavisRayM You'd have to create a new invitation

Reply:
@DavisRayM Now, I am thinking would it be better to create a new invitation one or update the revoked invitation to pending?

Reply:
@DavisRayM I get your point though, maybe resending a revoked invite should be considered valid